### ISSUE TITLE
How to control number of CPUs in mash

### ISSUE THREAD

**Issue by kazusyoshiishigaki** [stephenslab/mashr#71]
Hi,

I am using mash for a large eQTL dataset where I have more than 30 conditions. I am running mash using around 100K rows x 30 conditions. I am testing one job interactively in a computing node.
```
   data = mash_set_data(mat_b, mat_se) 
   m.ed = mash( data, U.ed ) 
```
I noticed that this process uses more than 20 CPUs. Is there any way to specifically control the number of CPUs when we use mash?

**Comment by pcarbo** [comment]
@kazusyoshiishigaki mash is using multithreaded matrix operations. So it would depend on the BLAS numerical library that R is using in your installation.

There is also an `mc.cores` argument to `calc_lik_matrix`, but the default is only 1 core.

**Comment by nicolerg** [comment]
Here's how I've been able to control the number of cores being used by `mash`, i.e. by telling BLAS how many cores to use, as @pcarbo mentioned (here, N CPUs = 12).
```
library(doParallel)
registerDoParallel(cores = 12)
library(RhpcBLASctl)
blas_set_num_threads(12)
```

**Comment by mural123** [comment]
How to use multiple cores with "mashr", without them being idle. I have been trying and it seems mashr dosent use all the cores. Mashr takes forever to complete the job and hit the maximum time limit permitted by our computing service to run the job and thus stop.

**Comment by gaow** [comment]
@mural123 which function in mashr are you referring to? Could you update to the current `master` branch or the current cran release for mashr and try again? We have improved openMP codes to resolve some compatibility issues with older versions of gcc / openMP. That might solve your problem.

**Comment by mural123** [comment]
Hi Dr. Gaow, I am refering to fitting the mash model. I did update mashr and tried the same.
m = mash(data.random, Ulist = c(U.ed,U.c), outputlevel = 1)
 - Computing 90000 x 8576 likelihood matrix.

**Comment by gaow** [comment]
@mural123 could you make sure you have openMP properly installed -- to do so, could you reinstall mashr and show us here the screen output as the package is being compiled and configured? We'll be able to tell you if your openMP has been configured during mashr install.

**Comment by stivenming** [comment]
I have encountered the same issue as @ravimural does. And here is the output of the mash installation process. 
<img width="761" alt="image" src="https://user-images.githubusercontent.com/45038348/226198497-7e24f2a0-c361-47d9-9f05-c7661bcba0be.png">
Furthermore, our compute cluster has an OpenMP environment, but still, mash is only using one core for calculations.
<img width="449" alt="image" src="https://user-images.githubusercontent.com/45038348/226198686-e5e9f30f-d9f3-4fff-a0a9-3e6adeb7c0d2.png">

**Comment by pcarbo** [comment]
@stivenming First, you should make sure `_OPENMP` is defined (see `mash.h`). 

Also note that not all mashr functions are multithreaded; only the functions with a `n_thread` argument can use >1 thread.

**Comment by stivenming** [comment]
@pcarbo Yes, `_OPENMP` is defined in `mash.h`, and all functions in `mash.cpp` have a `n_thread` parameter. However, the `mash()` function does not have a `n_thread` parameter in R console. Could you give me an example of how to enable multithreading? I'm a bit confused now. Thanks
